packages/next-swc/crates/next-build/src/next_build.rs

```diff
@@ -0,0 +1,588 @@
+                // Although this should perhaps be enforced by Turbopack semantics.
+                let all_client_assets: Vec<_> = all_assets_from_entries(client_chunks)
+                    .await?
+                    .iter()
+                    .map(|asset| {
+                        let client_root = client_root.clone();
+                        async move {
+                            Ok((
+                                asset.ident().path().await?.is_inside(&*client_root.await?),
+                                asset,
+                            ))
+                        }
+                    })
+                    .try_join()
+                    .await?
+                    .into_iter()
+                    .filter_map(|(is_inside, asset)| if is_inside { Some(*asset) } else { None })
+                    .collect();
+
+                Ok(Some((
+                    pathname,
+                    page_chunk.node_chunk,
+                    all_node_assets,
+                    client_chunks,
+                    all_client_assets,
+                )))
+            })
+            .try_join()
+            .await?
+            .into_iter()
+            .filter_map(|x| x)
+            .collect::<Vec<_>>();
+
+        {
+            let build_manifest_dir_path = build_manifest_path.parent().await?;
+            let pages_manifest_dir_path = pages_manifest_path.parent().await?;
+
+            let mut deduplicated_node_assets = HashMap::new();
+            let mut deduplicated_client_assets = HashMap::new();
+
+            // TODO(alexkirsz) We want all assets to emit them to the output directory, but
+            // we only want runtime assets in the manifest. Furthermore, the pages
+            // manifest (server) only wants a single runtime asset, so we need to
+            // bundle node assets somewhat.
+            for (pathname, node_chunk, all_node_assets, client_chunks, all_client_assets) in
+                page_chunks_and_url
+            {
+                tracing::debug!("pathname: {}", pathname.to_string(),);
+                tracing::debug!(
+                    "node chunk: {}",
+                    node_chunk.ident().path().to_string().await?
+                );
+                tracing::debug!(
+                    "client_chunks:\n{}",
+                    client_chunks
+                        .await?
+                        .iter()
+                        .map(|chunk| async move {
+                            Ok(format!("  - {}", chunk.ident().path().to_string().await?))
+                        })
+                        .try_join()
+                        .await?
+                        .join("\n")
+                );
+
+                // TODO(alexkirsz) Deduplication should not happen at this level, but
+                // right now we have chunks with the same path being generated
+                // from different entrypoints, and writing them multiple times causes
```

Comment:
Yes, but in this case I believe the assets are the same, but don't `Vc::resolve` to the same task. Otherwise, the graph traversal wouldn't have added them multiple times to the output. I think this is a bug, hence the comment.